### ISSUE TITLE
Special locale value for "any locale" or "unknown locale"

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#13526]
While discussing about #12673, @benharsh suggested that maybe we need a special value to say "any locale". The use case was where we are adding local indices to a distributed sparse domain. Currently the interface for that looks like:

```chapel
addBulk(indices, ..., addOn=nil:locale)
```

where the `nil` basically signals that user doesn't know where the indices should go. A specific value there could be more expressive.

Related: #7706

**Comment by e-kayrakli** [comment]
Quoting @benharsh's comment on #12673:

> What if we had a special locale value indicating "any locale", or maybe "unknown locale"? We have something sort of like this in NUMA with child locales called c_sublocid_any. In the context of our bulk-add method(s) could this special value indicate that the method should figure out the correct locales itself, and otherwise will add indices to the specified locale? Does that appeal to anyone?

**Comment by bradcray** [comment]
W.r.t. the status quo, I don't think that `nil:locale` should be supported...  I don't think the user should think about the locale as a class value, nor something that's nilable (there was another recent discussion that bumped against this similar topic, though I'd have to think harder to remember what it was).

Stylistically, `anywhere` seems like a nicer name for such a concept rather than `nil:locale` and symmetric with `here`.  The big question for me is whether it should be a reserved word or not (where my inclination is "not" if we can get away with that).

To make sure I'm understanding, what would the behavior be if that value was supplied?  Is the assumption that all of the indices will belong to a single locale but that we don't know which one it is?  Or that the set of indices may all belong to different locales?

**Comment by e-kayrakli** [comment]
@bradcray 

> To make sure I'm understanding, what would the behavior be if that value was supplied? Is the assumption that all of the indices will belong to a single locale but that we don't know which one it is? Or that the set of indices may all belong to different locales?

The latter. It simply means "I don't know. Go figure that out yourself". In that sense I see that `wherever` is a better fit in terms of what we mean, but I prefer `anywhere`, as well. 

Although such a special value (should we call it a "locality hint"?) sounds nice conceptually, I am also wondering if there is any place else other than `bulkAdd` where supporting it would be useful.

**Comment by e-kayrakli** [comment]
#15149 will make `locale` a record-wrapped type. As of now that PR adds a special value `nilLocale` that replaces the default value of the argument `addOn=nil:locale`. With that PR, we can introduce something like `anywhere` or `wherever`.

OTOH, I don't see this as part of the language stability effort. It is more of an advanced and somewhat obscure part of sparse domain interface, which itself is prone to some changes. So, I am leaning towards punting this decision for the next release. Although I am not sure how to do that from an implementation perspective. We can make these methods give a `--warn-unstable` warning when `nilLocale` is used. But then, what should be the default value for the argument? Maybe we can do a little better by adding a value `anyLocale` for now and use that. Note that, I don't expect the user code to pass that value to these methods. It is only used as a sentinel value for detecting that the user didn't pass anything to that argument.

**Comment by bradcray** [comment]
I would agree that it feels slightly late in the game to wrestle too much with sentinel locale values at this point and that anything we put in should be thought of as subject to change.  I'm open to tucking whatever sentinel values we need to keep things working on master for the time being if they come with that caveat and aren't advertised too broadly until they get more discussion and attention.

**Comment by e-kayrakli** [comment]
I am currently leaning towards adding a warning on the relevant method's documentation stating that `nilLocale` is a temporary workaround and possibly link this issue.
